From 316a8c56ab0327333865c6cabb390c82d3fdc7ab Mon Sep 17 00:00:00 2001
From: Ecco park <eccopark@broadcom.com>
Date: Mon, 11 Aug 2014 16:49:47 -0700
Subject: [PATCH 46/77] net: wireless: bcmdhd fixed RTT bugs

1) 16951581 Trying to run RTT with 16 configs reboots the phone
2) 16874783 rtt_sd values returned in rtt status don't seem correct
Signed-off-by: Ecco park <eccopark@broadcom.com>
---
 drivers/net/wireless/bcmdhd/dhd_rtt.c      |  7 ++++++-
 drivers/net/wireless/bcmdhd/wl_cfgvendor.c | 32 ++++++++++++++++++++++++++++--
 2 files changed, 36 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/dhd_rtt.c b/drivers/net/wireless/bcmdhd/dhd_rtt.c
index 167abf0fec22..b96bf1cbda75 100644
--- a/drivers/net/wireless/bcmdhd/dhd_rtt.c
+++ b/drivers/net/wireless/bcmdhd/dhd_rtt.c
@@ -83,6 +83,10 @@ chanspec_t
 dhd_rtt_convert_to_chspec(wifi_channel_info_t channel)
 {
 	int bw;
+	/* set witdh to 20MHZ for 2.4G HZ */
+	if (channel.center_freq >= 2400 && channel.center_freq <= 2500) {
+		channel.width = WIFI_CHAN_WIDTH_20;
+	}
 	switch (channel.width) {
 	case WIFI_CHAN_WIDTH_20:
 		bw = WL_CHANSPEC_BW_20;
@@ -424,7 +428,8 @@ dhd_rtt_convert_to_host(rtt_result_t *rtt_results, const wl_proxd_event_data_t*
 	else
 		DHD_RTT(("sigma:0\n"));
 
-	DHD_RTT(("rssi:%d validfrmcnt %d\n", rtt_results->avg_rssi, rtt_results->validfrmcnt));
+	DHD_RTT(("rssi:%d validfrmcnt %d, err_code : %d\n", rtt_results->avg_rssi,
+						rtt_results->validfrmcnt, evp->err_code));
 
 	switch (evp->err_code) {
 	case TOF_REASON_OK:
diff --git a/drivers/net/wireless/bcmdhd/wl_cfgvendor.c b/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
index a75225c19ef1..52f742b78e66 100644
--- a/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
+++ b/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
@@ -768,7 +768,7 @@ void wl_cfgvendor_rtt_evt(void *ctx, void *rtt_data)
 		rtt_report->tx_rate = rtt_result->tx_rate;
 		/* RTT */
 		rtt_report->rtt = rtt_result->meanrtt;
-		rtt_report->rtt_sd = rtt_result->sdrtt;
+		rtt_report->rtt_sd = rtt_result->sdrtt/10;
 		/* convert to centi meter */
 		if (rtt_result->distance != 0xffffffff)
 			rtt_report->distance = (rtt_result->distance >> 2) * 25;
@@ -791,7 +791,9 @@ static int wl_cfgvendor_rtt_set_config(struct wiphy *wiphy, struct wireless_dev
 	const struct nlattr *iter, *iter1, *iter2;
 	int8 eabuf[ETHER_ADDR_STR_LEN];
 	int8 chanbuf[CHANSPEC_STR_LEN];
+	int32 feature_set = 0;
 	struct bcm_cfg80211 *cfg = wiphy_priv(wiphy);
+	feature_set = dhd_dev_get_feature_set(bcmcfg_to_prmry_ndev(cfg));
 
 	WL_DBG(("In\n"));
 	err = dhd_dev_rtt_register_noti_callback(wdev->netdev, wdev, wl_cfgvendor_rtt_evt);
@@ -809,6 +811,7 @@ static int wl_cfgvendor_rtt_set_config(struct wiphy *wiphy, struct wireless_dev
 				WL_ERR(("exceed max target count : %d\n",
 					rtt_param.rtt_target_cnt));
 				err = BCME_RANGE;
+				goto exit;
 			}
 			break;
 		case RTT_ATTRIBUTE_TARGET_INFO:
@@ -818,13 +821,38 @@ static int wl_cfgvendor_rtt_set_config(struct wiphy *wiphy, struct wireless_dev
 					type = nla_type(iter2);
 					switch (type) {
 					case RTT_ATTRIBUTE_TARGET_MAC:
-						memcpy(&rtt_target->addr, nla_data(iter2), ETHER_ADDR_LEN);
+						memcpy(&rtt_target->addr, nla_data(iter2),
+							ETHER_ADDR_LEN);
 						break;
 					case RTT_ATTRIBUTE_TARGET_TYPE:
 						rtt_target->type = nla_get_u8(iter2);
+						if (!(feature_set & WIFI_FEATURE_D2D_RTT)) {
+							if (rtt_target->type == RTT_TWO_WAY ||
+								rtt_target->type == RTT_INVALID) {
+								WL_ERR(("doesn't support RTT type : %d\n",
+									rtt_target->type));
+								err = -EINVAL;
+								goto exit;
+							} else if (rtt_target->type == RTT_AUTO) {
+								rtt_target->type = RTT_ONE_WAY;
+							}
+						} else {
+							if (rtt_target->type == RTT_INVALID) {
+								WL_ERR(("doesn't support RTT type : %d\n",
+									rtt_target->type));
+								err = -EINVAL;
+								goto exit;
+							}
+						}
 						break;
 					case RTT_ATTRIBUTE_TARGET_PEER:
 						rtt_target->peer= nla_get_u8(iter2);
+						if (rtt_target->peer != RTT_PEER_AP) {
+							WL_ERR(("doesn't support peer type : %d\n",
+								rtt_target->peer));
+							err = -EINVAL;
+							goto exit;
+						}
 						break;
 					case RTT_ATTRIBUTE_TARGET_CHAN:
 						memcpy(&rtt_target->channel, nla_data(iter2),
-- 
2.0.3

